### PR TITLE
View Data Definitions in Define

### PR DESCRIPTION
**Pull request by NathanielScottStevens**

### CODE REVIEW

**Comment by jmorris-pillar** [comment]
Please squash and merge when merging this pr

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,13 @@
...
+COPY apps/service_define/assets/ /app
+WORKDIR /app/
+RUN npm install \
+  && npm run release
+
 FROM bitwalker/alpine-elixir-phoenix:1.9.4 as build
 COPY . /opt/app
-WORKDIR /opt/app
+WORKDIR /opt/app/
+COPY --from=npm_builder /app/dist/ apps/service_define/priv/static
```
Comment (jmorris-pillar) [inline Dockerfile:10]: Why do we copy from dist to priv static?  What normally takes care of this in the release process that we cant just put this in a dist folder?

File: apps/service_define/assets/src/component/presentation/data-definition/DataDefinition.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            struct_module_name: "Elixir.Extractor",
+            args: [
+                {
+                    key: "steps",
+                    type: ["list", PrimitiveArgumentType.module],
+                    value: [
+                        {
+                            struct_module_name: "Extract.Http.Get",
+                            args: [
+                                { key: "url", type: PrimitiveArgumentType.string, value: "http://www.source.com/the-data.csv"}
```
Comment (jmorris-pillar) [inline apps/service_define/assets/src/component/presentation/data-definition/DataDefinition.stories.tsx:21]: Can we use example.com or something with .example as the tld just to be good denizens?

Comment (jmorris-pillar) [reply apps/service_define/assets/src/component/presentation/data-definition/DataDefinition.stories.tsx:21]: The top-level domain example is explicitly intended to be used in documentation or other technical writing, when domain names are presented as examples in usage or presentation of concepts of the Domain Name System or the Internet.

File: apps/service_define/config/dev.exs [hunk 3]
```diff
@@ -1,5 +1,20 @@
 use Mix.Config
 
+config :service_define, Define.Application,
+  kafka_endpoints: nil,
```
Comment (jmorris-pillar) [inline apps/service_define/config/dev.exs:4]: Do you need this line?

File: apps/service_define/config/dev.exs [hunk 4]
```diff
@@ -14,9 +29,7 @@ config :service_define, DefineWeb.Endpoint,
       "../priv/static/",
       cd: Path.expand("../assets", __DIR__)
     ]
-  ]
-
-config :service_define, DefineWeb.Endpoint,
+  ],
   live_reload: [
```
Comment (jmorris-pillar) [inline apps/service_define/config/dev.exs:33]: Did you mean to delete this line that now puts live reload under the main app and not under the endpoint?

File: apps/service_define/dev_scripts/send_events.exs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    dataset_id: id,
+    subset_id: "default",
+    destination: "success",
+    dictionary: [
+        Dictionary.Type.String.new!(name: "letter")
+    ],
+    steps: [
+        Extract.Http.Get.new!(
+        url: "http://localhost/file.csv",
+        headers: %{"content-length" => "5"}
```
Comment (jmorris-pillar) [inline apps/service_define/dev_scripts/send_events.exs:15]: Why do we need to tell the content length on the request as a header?

File: apps/service_define/mix.exs [hunk 6]
```diff
@@ -29,13 +30,21 @@ defmodule Define.MixProject do
...
+      {:extractor, in_umbrella: true},
       {:gettext, "~> 0.11"},
       {:jason, "~> 1.0"},
       {:metrics_reporter, in_umbrella: true},
       {:phoenix, "~> 1.4.15"},
       {:phoenix_pubsub, "~> 1.1"},
       {:plug_cowboy, "~> 2.0"},
       {:properties, in_umbrella: true},
+      # This should eventually be transformer once that has been created (like extractor)
+      {:transform_wkt, in_umbrella: true}
```
Comment (jmorris-pillar) [inline apps/service_define/mix.exs:47]: comma

File: helm/presets/local.yaml [hunk 7]
```diff
@@ -1,3 +1,6 @@
+define:
+  service:
+    type: NodePort
```
Comment (jmorris-pillar) [inline helm/presets/local.yaml:3]: Should this be a different service type than acquire?

File: helm/templates/define/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+              protocol: TCP
+          env:
+            - name: KAFKA_ENDPOINTS
+              value: {{ template "hindsight.fullname" . }}-kafka-bootstrap:9092
+            - name: SECRET_KEY_BASE
+              valueFrom:
+                secretKeyRef:
+                  name: {{ template "hindsight.fullname" . }}-define-key-base
+                  key: secret_key_base
+            - name: REDIS_HOST
```
Comment (jmorris-pillar) [inline helm/templates/define/deployment.yaml:47]: One of you // kenny will need to make sure this gets reconciled to postgres

Comment (jmorris-pillar) [reply helm/templates/define/deployment.yaml:47]: i.e. the race is on.

File: helm/presets/local.yaml [hunk 9]
```diff
@@ -1,3 +1,6 @@
+define:
```
Comment (jmorris-pillar) [inline helm/presets/local.yaml:1]: Are you also going ot need to have changes to aws.yaml?  We probably need this configured behind the ingress controller which I can help you with.  I'm okay if we do that in seperate pull requests to keep this one from getting longer.

File: .dockerignore [hunk 10]
```diff
@@ -4,4 +4,7 @@ deps/
 .elixir_ls/
 .git/
 .github/
-helm/
\ No newline at end of file
+helm/
+apps/service_define/assets/node_modules/
+apps/service_define/assets/.cache/
+apps/service_define/assets/build/
```
Comment (jeffgrunewald) [inline .dockerignore:10]: i take it there _are_ assets we want to upload so you can't just omit the entire `apps/service_define/assets/` dir?

File: apps/definition/lib/definition/schema/type.ex [hunk 11]
```diff
@@ -44,7 +44,29 @@ defmodule Definition.Schema.Type do
...
+  end
+
+  @spec atom() :: spec
+  def atom() do
+    spec(is_atom())
+  end
+
+  @spec boolean() :: spec
+  def boolean() do
+    spec(is_boolean())
```
Comment (jeffgrunewald) [inline apps/definition/lib/definition/schema/type.ex:67]: i'll leave it to @bbalser or @jdenen since they've been deeper in Norm for longer: is it preferred to just directly call `spec(predicate_func())` directly when you're invoking Norm, or is it preferable to wrap these `spec/1` calls?

File: apps/definition_dictionary/lib/dictionary/type/string.ex [hunk 12]
```diff
@@ -2,6 +2,12 @@ defmodule Dictionary.Type.String do
   use Definition, schema: Dictionary.Type.String.V1
   use Dictionary.JsonEncoder
 
+  @type t :: %__MODULE__{
+    version: integer,
+    name: String.t(),
+    description: String.t()
+  }
```
Comment (jeffgrunewald) [inline apps/definition_dictionary/lib/dictionary/type/string.ex:9]: many type definitions lost their lives to bring us that big source/destination refactor... 😞

File: apps/service_define/config/dev.exs (hunk 3 above)
Comment (jeffgrunewald) [reply apps/service_define/config/dev.exs:4]: we shouldn't; it's not explicitly defined in `config.exs` so it should default to `nil`

File: apps/service_define/config/dev.exs [hunk 13]
```diff
@@ -1,5 +1,20 @@
...
+    driver: [
+      module: Brook.Driver.Default,
+      init_arg: []
+    ],
+    handlers: [Define.Event.Handler],
+    storage: [
+      module: Brook.Storage.Ets,
+      init_arg: []
+    ],
+    dispatcher: Brook.Dispatcher.Noop
```
Comment (jeffgrunewald) [inline apps/service_define/config/dev.exs:15]: Since we're supplying these values "at runtime" in the tests by doing overrides using the `Properties` macro, are we including these here for running the application locally while doing development of the UI? if so, you should be able to remove the same settings in `test.exs` and supply the values in your specific tests by the `temp_env` module in the `:testing` umbrella app.

File: apps/service_define/lib/define/event/store.ex [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  def update_definition(%Load{} = data) do
+    update_definition_field(data, :load, &to_load_view/1)
+  end
+
+  def update_definition(%Extract{} = data) do
+    update_definition_field(data, :extract, &to_extract_view/1)
+  end
+
+  def update_definition(data) do
+    Logger.error("Got unexpected data definition update: #{inspect(data)}")
```
Comment (jeffgrunewald) [inline apps/service_define/lib/define/event/store.ex:23]:
```suggestion
    Logger.error(fn -> "Got unexpected data definition update: #{inspect(data)}" end)
```

File: apps/service_define/lib/define/event/store.ex [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  defp update_definition_field(data, key, to_view_converter) do
+    get_or_create(data)
+    |> Map.put(key, to_view_converter.(data))
+    |> persist()
+  rescue
+    err ->
+      Logger.error("Unable to process transform event: #{inspect(err)}")
```
Comment (jeffgrunewald) [inline apps/service_define/lib/define/event/store.ex:32]:
```suggestion
      Logger.error(fn -> "Unable to process transform event: #{inspect(err)}" end)
```

File: helm/presets/local.yaml (hunk 7 above)
Comment (jeffgrunewald) [reply helm/presets/local.yaml:3]: this should be a `ClusterIP` type and include a path in the ingress manifest to reach Define directly from the web on 80.

File: .dockerignore (hunk 10 above)
Comment (nbeach) [reply .dockerignore:10]: This is correct. assets/ contains all the UI code, and these a just a few folders that get generated that we don't want.

File: Dockerfile (hunk 1 above)
Comment (nbeach) [reply Dockerfile:10]: Not sure what you mean by "just put this in a dist folder"? priv/static/ is where phoenix scoops up the static assets when released, so we first have to use a node docker image to build the UI, then copy those built files to the proper folder in the elixir image before the release is run. Is there a simplification we're missing here?

File: apps/service_define/dev_scripts/send_events.exs (hunk 5 above)
Comment (NathanielScottStevens) [reply apps/service_define/dev_scripts/send_events.exs:15]: We don't, it was just what we thought of first. Would you prefer we use something else?

File: apps/definition/lib/definition/schema/type.ex (hunk 11 above)
Comment (NathanielScottStevens) [reply apps/definition/lib/definition/schema/type.ex:67]: Yeah, we were curious about this too. We were trying to follow precedent but may have understood that precedent incorrectly.

File: apps/service_define/config/dev.exs (hunk 13 above)
Comment (NathanielScottStevens) [reply apps/service_define/config/dev.exs:15]: I don't really know anything about the testing umbrella app. Why would we put our configuration there instead of with the apps?

File: apps/service_define/assets/src/component/presentation/data-definition/DataDefinition.stories.tsx (hunk 2 above)
Comment (NathanielScottStevens) [reply apps/service_define/assets/src/component/presentation/data-definition/DataDefinition.stories.tsx:21]: I almost suggested this while pairing and then thought "no, that's a little too pedantic". I'll put it in for you though Jessie.

File: apps/service_define/config/dev.exs (hunk 4 above)
Comment (NathanielScottStevens) [reply apps/service_define/config/dev.exs:33]: `live_reload` is still under the endpoint.  This diff is just really confusing I think.

File: helm/presets/local.yaml (hunk 9 above)
Comment (NathanielScottStevens) [reply helm/presets/local.yaml:1]: I will add that now. Although I'm not sure what needs to be done with the ingress controller. I'll need your help there for sure.

Comment (jeffgrunewald) [reply helm/presets/local.yaml:1]: Nothing needs done to the ingress _controller_ but you need to add another path to `Values.ingress.paths` in the root values.yaml file and change the define service to a ClusterIP (not required but more restricted than NodePort and obviates the need for NodePort)

File: apps/service_define/config/dev.exs (hunk 13 above)
Comment (jeffgrunewald) [reply apps/service_define/config/dev.exs:14]: Testing gives you a module called `Temp.Env` that lets you set app env configs in a specific test to avoid blanket configurations across the test environment and keeps the `test.exs` and `integration.exs` files much cleaner. It takes the app env stuff that we're setting via the Properties library (also umbrella) and configuring them at runtime, both defaulting to prod settings and overriding only as needed in the tests that require it.

Look at the service_persist's `lib/event/handler.ex` and its associated test to see how we're defining configuration and then overriding it as needed in a test.

File: apps/service_define/mix.exs [hunk 16]
```diff
@@ -11,6 +11,7 @@ defmodule Define.MixProject do
       lockfile: "../../mix.lock",
       elixir: "~> 1.5",
       elixirc_paths: elixirc_paths(Mix.env()),
+      elixirc_options: [debug_info: true],
```
Comment (jeffgrunewald) [inline apps/service_define/mix.exs:14]: is this something we want turned on permanently?

File: apps/service_define/config/dev.exs (hunk 13 above)
Comment (NathanielScottStevens) [reply apps/service_define/config/dev.exs:14]: It seems simpler to just to provide them through the config for now until we have a reason to override them in a test.

File: apps/service_define/lib/define/event/store.ex [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  defp update_definition_field(data, key, to_view_converter) do
+    get_or_create(data)
+    |> Map.put(key, to_view_converter.(data))
+    |> persist()
+  rescue
+    err ->
+      Logger.error(fn -> "Unable to process transform event: #{inspect(err)}" end)
+      discard()
```
Comment (jdenen) [inline apps/service_define/lib/define/event/store.ex:33]:
```suggestion
      :discard
```
I think we've taken the abstraction too far.

File: apps/service_define/lib/define/event/handler.ex [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+    Define.Event.Store.update_definition(extract)
+  end
+
+  def handle_event(%Brook.Event{type: transform_define(), data: %Transform{} = transform}) do
+    Define.Event.Store.update_definition(transform)
+  end
+
+  def handle_event(%Brook.Event{type: load_start(), data: %Load{} = load}) do
+    Define.Event.Store.update_definition(load)
+  end
```
Comment (jdenen) [inline apps/service_define/lib/define/event/handler.ex:17]:
```suggestion
  @events [extract_start(), transform_define(), load_start()]
  
  def handle_event(%Brook.Event{type: event, data: data}) when event in @events do
    Define.Event.Store.update_definition(data)
  end
```
Just seems like a place to DRY up.

File: apps/service_define/lib/define/event/store.ex [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    update_definition_field(data, :load, &to_load_view/1)
+  end
+
+  def update_definition(%Extract{} = data) do
+    update_definition_field(data, :extract, &to_extract_view/1)
+  end
+
+  def update_definition(data) do
+    Logger.error(fn -> "Got unexpected data definition update: #{inspect(data)}" end)
+  end
```
Comment (jdenen) [inline apps/service_define/lib/define/event/store.ex:24]: What's the spec of this function? Why is an unexpected input returning the same result as a good input?

File: apps/service_define/lib/define/event/store.ex [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+  defp update_definition_field(data, key, to_view_converter) do
+    get_or_create(data)
+    |> Map.put(key, to_view_converter.(data))
+    |> persist()
+  rescue
```
Comment (jdenen) [inline apps/service_define/lib/define/event/store.ex:30]: Not a big fan of using rescue like this. Can we refactor to use `:ok/:error` tuples throughout?

File: apps/service_define/lib/define/event/store.ex (hunk 19 above)
Comment (jdenen) [reply apps/service_define/lib/define/event/store.ex:24]: See my question about `:ok/:error` below. Same deal feels like it should apply here.

File: .dockerignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (jdenen) [reply .dockerignore:10]: I honestly don't know, so feel free to ignore.

File: apps/service_define/lib/define/typespec_analysis.ex [hunk 21]
```diff
@@ -0,0 +1,68 @@
+defmodule Define.TypespecAnalysis do
```
Comment (jdenen) [inline apps/service_define/lib/define/typespec_analysis.ex:1]: Definitely document this.

File: apps/service_define/lib/define/typespec_analysis.ex [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+  import Logger
+
+  def find() do
```
Comment (jdenen) [inline apps/service_define/lib/define/typespec_analysis.ex:4]:
```suggestion
  def find do
```
Also add a spec, please.

File: apps/service_define/lib/define/typespec_analysis.ex [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+
+      :error ->
+        Logger.error("Unable to get types for #{module}")
+        raise(ArgumentError, "Unable to get types for #{module}")
+    end
+  end
+
+  defp extract_fields([]) do
+    []
+  end
```
Comment (jdenen) [inline apps/service_define/lib/define/typespec_analysis.ex:33]:
```suggestion
  defp extract_fields([]), do: []
```

Comment (jdenen) [reply apps/service_define/lib/define/typespec_analysis.ex:33]: I'm getting real nit-picky, I know. Sorry.

File: apps/service_define/lib/define/typespec_analysis.ex [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+  defp extract_field_type({:type, _, _, [{_, _, name}, {_, _, type}]}), do: {name, type, nil}
+
+  defp extract_field_type({:type, _, _, [{_, _, name}, {_, _, type, sub_typespec}]}),
+    do: {name, type, sub_typespec}
+
+  defp to_simple_type({name, :list, [{_, _, [{_, _, String}, _, _]}]}),
+    do: {to_string(name), {"list", "string"}}
+
+  defp to_simple_type({name, :list, [{_, _, type, _}]}),
+    do: {to_string(name), {"list", to_string(type)}}
```
Comment (jdenen) [inline apps/service_define/lib/define/typespec_analysis.ex:52]: Oof, these are hard to parse. Can we make them proper `do/end` style functions. That'll at least break what parts I'm supposed to care about from the ones I'm not.

File: apps/service_define/lib/define/typespec_analysis.ex [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+  defp to_simple_type({name, typespec, _}) do
+    type =
+      case typespec do
+        [{_, _, String}, _, _] -> "string"
+        [{_, _, Dictionary}, _, _] -> "dictionary"
+        _ -> typespec
+      end
+
+    {to_string(name), to_string(type)}
+  end
```
Comment (jdenen) [inline apps/service_define/lib/define/typespec_analysis.ex:63]: I think this could be simplified by pulling parsing logic out of the function head and into a private function:

```elixir
defp to_simple_type({name, typespec, details}) do
  type = 
    case typespec do
      :list -> {"list", list_item_simple_type(details)}
      [{_, _, String} | _] -> "string"
      [{_, _, Dictionary} | _] -> "dictionary"
      _ -> to_string(typespec)
    end

  {to_string(name), type}
end

defp list_item_simple_type([{_, _, [{_, _, String} | _]}]), do: "string"
defp list_item_simple_type([{_, _, type, _}]), do: to_string(type)
```

Comment (jdenen) [reply apps/service_define/lib/define/typespec_analysis.ex:63]: Hmm. Even after writing that out I believe this should be simpler. Maybe we should parse the `{name, typespec, details}` tuple into a struct and have helper functions that return field and simple types.

It feels complicated to parse field and simple types from nested and convoluted tuples/lists.

File: apps/service_define/lib/define/view_state.ex [hunk 26]
```diff
@@ -20,9 +23,8 @@ defmodule Define.ViewState do
   end
 
   @impl true
-  def handle_call({:event, type, payload}, _from, state) do
-    new_state = update_state(state, type, payload)
-    {:reply, new_state, new_state}
+  def handle_call({:event, _type, _payload}, _from, state) do
+    {:reply, state, state}
```
Comment (jdenen) [inline apps/service_define/lib/define/view_state.ex:27]: I assume we continue to keep this module for what we know to be coming? We might want to document that for now.

File: apps/service_define/test/define/typespec_analysis_test.exs [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+      "map_type" => "map",
+      "dictionary_type" => "dictionary"
+    }
+
+    assert expected == TypespecAnalysis.get_types(TypespecAnalysisTestModule)
+  end
+
+  test "returns empty map when a typespec is not defined" do
+    assert %{} == TypespecAnalysis.get_types(TypespecAnalysisTestModuleNoSpec)
+  end
```
Comment (jdenen) [inline apps/service_define/test/define/typespec_analysis_test.exs:22]:
```suggestion
  describe "get_types/1" do
    test "returns all types from the typespec" do
      expected = %{
        "integer_type" => "integer",
        "string_type" => "string",
        "atom_type" => "atom",
        "float_type" => "float",
        "string_list_type" => {"list", "string"},
        "integer_list_type" => {"list", "integer"},
        "map_type" => "map",
        "dictionary_type" => "dictionary"
      }

      assert expected == TypespecAnalysis.get_types(TypespecAnalysisTestModule)
    end

    test "returns empty map when a typespec is not defined" do
      assert %{} == TypespecAnalysis.get_types(TypespecAnalysisTestModuleNoSpec)
    end
  end
```

This is more or less the pattern we're following for grouping tests around a single function.

File: Dockerfile (hunk 1 above)
Comment (jmorris-pillar) [reply Dockerfile:10]: Probably just somethign I'm missing locally.  I figure if you ran it locally you run something like npm run release inside the assets folder and dont have to then copy them to priv/static, but im guessing theres some nuance with distillery we're chasing?

File: apps/service_define/assets/package.json [hunk 28]
```diff
@@ -8,7 +8,7 @@
     "compile": "tsc --noEmit",
     "ci": "npm run compile && npm run test:once && npm run lint",
     "storybook": "start-storybook -s ./public",
-    "deploy": "parcel build src/index.html --out-dir ../priv/static/",
+    "release": "parcel build src/index.html --out-dir ./dist",
```
Comment (jmorris-pillar) [inline apps/service_define/assets/package.json:11]: It is possible this is the missing piece to my dockerfile question.

File: apps/service_define/config/dev.exs (hunk 4 above)
Comment (jmorris-pillar) [reply apps/service_define/config/dev.exs:32]: sure is, thanks lol

File: apps/service_define/lib/define/event/handler.ex (hunk 18 above)
Comment (jmorris-pillar) [reply apps/service_define/lib/define/event/handler.ex:17]: Do you lose a secondary check that those event types have the right structs though?  Does that matter?

Comment (jdenen) [reply apps/service_define/lib/define/event/handler.ex:17]: Take a look at `apps/definition_events/lib/events.ex`.

This macro creates send_* functions for the events that ensure event type and event struct are locked together.

File: apps/service_define/config/dev.exs (hunk 13 above)
Comment (NathanielScottStevens) [reply apps/service_define/config/dev.exs:14]: So that we can run the app locally without any dependencies. I guess we could add it to the top level umbrella base config but that seems weird to me.

File: apps/definition/lib/definition/schema/type.ex [hunk 29]
```diff
@@ -44,7 +44,29 @@ defmodule Definition.Schema.Type do
     end)
   end
 
-  @spec access_path() :: spec
+  @spec map() :: spec
+  def map() do
+    spec(fn
+      %{__struct__: _} -> false
+      x -> is_map(x)
```
[the remark this answers is not in the document]
Comment (NathanielScottStevens) [reply apps/definition/lib/definition/schema/type.ex:51]: Yes, we handle structs and maps differently when we generate the UI

File: apps/service_define/lib/define/event/store.ex (hunk 17 above)
Comment (NathanielScottStevens) [reply apps/service_define/lib/define/event/store.ex:33]: Yeah, I don't care for this either.

File: apps/service_define/lib/define/event/store.ex (hunk 20 above)
Comment (NathanielScottStevens) [reply apps/service_define/lib/define/event/store.ex:30]: I originally took us down that path but then Nick pointed out that we were adding a lot of extra work/verbosity by doing so which I found myself agree with. I forget all the particulars about why it got so gross at this point though.

File: apps/service_define/mix.exs (hunk 16 above)
Comment (NathanielScottStevens) [reply apps/service_define/mix.exs:14]: No, good catch

File: apps/service_define/lib/define/event/store.ex (hunk 19 above)
Comment (NathanielScottStevens) [reply apps/service_define/lib/define/event/store.ex:24]: I'm not quite sure I'm following what you're thinking here. By an unexpected input do you mean an unexpected event? `update_definition_field` ends up storing the event in the view state whereas the catch all does not. Are you wanting us to bring the storing in the viewstate up to these top level functions to make that more apparent?

Comment (jdenen) [reply apps/service_define/lib/define/event/store.ex:25]: `Logger.error/1` returns `:ok` was what I was getting at.

If a valid definition is passed to `update_definition` but fails to persist, we log an error and return `:discard`.

If an invalid definition is passed to `update_definition`, we log an error and return `:ok`.

File: apps/service_define/lib/define/event/store.ex (hunk 20 above)
Comment (jdenen) [reply apps/service_define/lib/define/event/store.ex:31]: So we're going to use exceptions for control flow because of verbosity?

File: helm/presets/aws.yaml [hunk 30]
```diff
@@ -101,6 +104,14 @@ presto:
 cert_manager:
   enabled: true
 
+global:
```
Comment (jmorris-pillar) [inline helm/presets/aws.yaml:107]: Do you need to add this.  I think this reverts a change jeff recently made?

File: helm/presets/local.yaml [hunk 31]
```diff
@@ -66,7 +69,16 @@ presto:
       annotations:
         nginx.ingress.kubernetes.io/use-regex: "true"
       hosts: [storage.hindsight.local/minio.*]
-
+global:
```
Comment (jmorris-pillar) [inline helm/presets/local.yaml:72]: why do we need to add this?

File: helm/presets/aws.yaml (hunk 30 above)
Comment (NathanielScottStevens) [reply helm/presets/aws.yaml:107]: I'm guessing we just messed up the merge
